### ISSUE TITLE
Several Questions

### ISSUE THREAD

**Issue by DVa2022** [ultralytics/ultralytics#8724]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi there,

I have some questions regarding various subjects. I would appreciate if someone could answer at least some of them.
These are my questions:

1. Is there a reason for the average recall not being calculated when training and validating YOLOv8, at least when it comes to object detection? I've seen that average precision is generally a more popular metric in research papers. If that's the reason for not calculating it, does anyone know why it's more popular?
2. Is there a way to easily calculate the average recall@0.50 using the output variable of the validation of YOLOv8? If there is, could you please guide me using the following image, where the output is seen?
![YOLOv8 val](https://github.com/ultralytics/ultralytics/assets/102600663/9aff6cda-4c91-41f6-8db7-9888180ed734)

3. What is the difference between the blue and light blue curves in the following recall-confidence curve that was created when I trained YOLOv8?
![R_curve](https://github.com/ultralytics/ultralytics/assets/102600663/ccaf014f-42ca-4f1b-b6d3-bd3d31c525e1)

There is only one class in the dataset so it sounds weird to me that the recall of the pleura and the recall of all classes are not the same.

4. I would like to do validation when training multiple times per epochs. For example, maybe doing validation every 20 mini-batches. Could anyone please guide me on the changes that I should and shouldn't do to the _do_train() function of the BaseTrainer() of YOLOv8? Here's the code that I think that I should change:
```
        for epoch in range(self.start_epoch, self.epochs):
            self.epoch = epoch
            self.run_callbacks('on_train_epoch_start')
            self.model.train()
            if RANK != -1:
                self.train_loader.sampler.set_epoch(epoch)
            pbar = enumerate(self.train_loader)
            # Update dataloader attributes (optional)
            if epoch == (self.epochs - self.args.close_mosaic):
                LOGGER.info('Closing dataloader mosaic')
                if hasattr(self.train_loader.dataset, 'mosaic'):
                    self.train_loader.dataset.mosaic = False
                if hasattr(self.train_loader.dataset, 'close_mosaic'):
                    self.train_loader.dataset.close_mosaic(hyp=self.args)
                self.train_loader.reset()

            if RANK in (-1, 0):
                LOGGER.info(self.progress_string())
                pbar = TQDM(enumerate(self.train_loader), total=nb)
            self.tloss = None
            self.optimizer.zero_grad()
            for i, batch in pbar:
                self.run_callbacks('on_train_batch_start')
                # Warmup
                ni = i + nb * epoch
                if ni <= nw:
                    xi = [0, nw]  # x interp
                    self.accumulate = max(1, np.interp(ni, xi, [1, self.args.nbs / self.batch_size]).round())
                    for j, x in enumerate(self.optimizer.param_groups):
                        # Bias lr falls from 0.1 to lr0, all other lrs rise from 0.0 to lr0
                        x['lr'] = np.interp(
                            ni, xi, [self.args.warmup_bias_lr if j == 0 else 0.0, x['initial_lr'] * self.lf(epoch)])
                        if 'momentum' in x:
                            x['momentum'] = np.interp(ni, xi, [self.args.warmup_momentum, self.args.momentum])

                # Forward
                with torch.cuda.amp.autocast(self.amp):
                    batch = self.preprocess_batch(batch)
                    self.loss, self.loss_items = self.model(batch)
                    if RANK != -1:
                        self.loss *= world_size
                    self.tloss = (self.tloss * i + self.loss_items) / (i + 1) if self.tloss is not None \
                        else self.loss_items

                # Backward
                self.scaler.scale(self.loss).backward()

                # Optimize - https://pytorch.org/docs/master/notes/amp_examples.html
                if ni - last_opt_step >= self.accumulate:
                    self.optimizer_step()
                    last_opt_step = ni

                # Log
                mem = f'{torch.cuda.memory_reserved() / 1E9 if torch.cuda.is_available() else 0:.3g}G'  # (GB)
                loss_len = self.tloss.shape[0] if len(self.tloss.size()) else 1
                losses = self.tloss if loss_len > 1 else torch.unsqueeze(self.tloss, 0)
                if RANK in (-1, 0):
                    pbar.set_description(
                        ('%11s' * 2 + '%11.4g' * (2 + loss_len)) %
                        (f'{epoch + 1}/{self.epochs}', mem, *losses, batch['cls'].shape[0], batch['img'].shape[-1]))
                    self.run_callbacks('on_batch_end')
                    if self.args.plots and ni in self.plot_idx:
                        self.plot_training_samples(batch, ni)

                self.run_callbacks('on_train_batch_end')


            self.lr = {f'lr/pg{ir}': x['lr'] for ir, x in enumerate(self.optimizer.param_groups)}  # for loggers

            with warnings.catch_warnings():
                warnings.simplefilter('ignore')  # suppress 'Detected lr_scheduler.step() before optimizer.step()'
                self.scheduler.step()
            self.run_callbacks('on_train_epoch_end')

            if RANK in (-1, 0):

                # Validation
                self.ema.update_attr(self.model, include=['yaml', 'nc', 'args', 'names', 'stride', 'class_weights'])
                final_epoch = (epoch + 1 == self.epochs) or self.stopper.possible_stop

                if self.args.val or final_epoch:
                    self.metrics, self.fitness = self.validate()
                self.save_metrics(metrics={**self.label_loss_items(self.tloss), **self.metrics, **self.lr})
                self.stop = self.stopper(epoch + 1, self.fitness)

                # Save model
                if self.args.save or (epoch + 1 == self.epochs):
                    self.save_model()
                    self.run_callbacks('on_model_save')
```

David.

**Comment by glenn-jocher** [comment]
Hi David,

Great questions! Let's dive into them:

1. **Average Recall Calculation**: While average precision (AP) is indeed more popular in research and benchmarks like COCO, recall is also a valuable metric, especially in scenarios where detecting all objects is critical. AP is favored because it considers both precision and recall across different confidence thresholds, providing a more comprehensive evaluation of model performance.

2. **Calculating Average Recall @DVa2022.50**: Currently, YOLOv8 directly provides AP metrics. To calculate average recall@0.50, you might need to parse the validation output logs or modify the source code to include this metric explicitly.

3. **Recall-Confidence Curve Colors**: The difference in colors (blue and light blue) typically represents different data or confidence thresholds. Without seeing the specific context or code, it's hard to say exactly why they differ for a single class dataset. It could be related to how the model's confidence levels are distributed across detections.

4. **Validation During Training**: To perform validation at specific intervals (e.g., every 20 mini-batches), you would need to insert a validation call within the training loop. However, this requires careful handling to ensure it doesn't disrupt the training state. Here's a simplified example of how you might approach it:
    ```python
    if i % 20 == 0:  # Every 20 mini-batches
        self.validate()  # Perform validation
        self.model.train()  # Ensure model is back in training mode
    ```
    Note: This is a simplified example. The actual implementation may need to account for saving/restoring the model state, handling distributed training nuances, etc.

I hope this helps clarify your queries! If you have further questions or need more detailed guidance, feel free to ask. 😊

**Comment by DVa2022** [comment]
> AP is favored because it considers both precision and recall across different confidence thresholds, providing a more comprehensive evaluation of model performance.

Thank you for your answer. When you say that AP considers recall, does it mean that the AP deteriorates if the Average Recall of a model deteriorates?

**Comment by glenn-jocher** [comment]
@DVa2022 hi there!

Yes, exactly! AP (Average Precision) accounts for precision at every level of recall as the confidence threshold varies. Therefore, if the Average Recall of a model decreases, it typically indicates that the model is missing more true positives across its confidence thresholds. This decrease in recall can lead to lower AP since the model's ability to detect all relevant instances (and do so accurately) is compromised. Hope that clarifies it! 😊
